File: apps/python-tictactoe/game.py
"""Tic-Tac-Toe game logic."""

from dataclasses import dataclass, field
from enum import Enum
from typing import Optional
import uuid


class Player(str, Enum):
    X = "X"
    O = "O"


class Status(str, Enum):
    IN_PROGRESS = "in_progress"
    X_WINS = "x_wins"
    O_WINS = "o_wins"
    DRAW = "draw"
    X_FORFEITS = "x_forfeits"
    O_FORFEITS = "o_forfeits"


class NotYourTurn(Exception):
    """Raised when a player tries to move out of turn."""
    pass


class CellOccupied(Exception):
    """Raised when a player tries to move to an occupied cell."""
    pass


class InvalidPosition(Exception):
    """Raised when a move is outside board boundaries."""
    pass


class GameOver(Exception):
    """Raised when a move is attempted after the game has ended."""
    pass


@dataclass
class Game:
    """Represents a tic-tac-toe game."""

    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    board: list[str] = field(default_factory=lambda: [""] * 9)
    current_player: Player = Player.X
    status: Status = Status.IN_PROGRESS
    winner: Optional[Player] = None

    def move(self, row: int, col: int, player: Player) -> None:
        """Make a move on the board."""
        # Check if game is over
        if self.status != Status.IN_PROGRESS:
            raise GameOver()

        # Validate position
        if row < 0 or row > 2 or col < 0 or col > 2:
            raise InvalidPosition()

        # Check turn
        if player != self.current_player:
            raise NotYourTurn()

        # Calculate board index
        index = row * 3 + col

        # Check if cell is occupied
        if self.board[index] != "":
            raise CellOccupied()

        # Place the mark
        self.board[index] = player.value

        # Check for win
        if self._check_win(player):
            self.status = Status.X_WINS if player == Player.X else Status.O_WINS
            self.winner = player
            return

        # Check for draw
        if self._is_board_full():
            self.status = Status.DRAW
            return

        # Switch player
        self.current_player = Player.O if self.current_player == Player.X else Player.X

    def forfeit(self, player: Player) -> None:
        """Current player forfeits the game."""
        # Check if game is over
        if self.status != Status.IN_PROGRESS:
            raise GameOver()

        # Check turn
        if player != self.current_player:
            raise NotYourTurn()

        # Set forfeit status and winner
        if player == Player.X:
            self.status = Status.X_FORFEITS
            self.winner = Player.O
        else:
            self.status = Status.O_FORFEITS
            self.winner = Player.X

    def _check_win(self, player: Player) -> bool:
        """Check if the given player has won."""
        mark = player.value

        # Winning combinations (rows, columns, diagonals)
        lines = [
            # Rows
            [0, 1, 2], [3, 4, 5], [6, 7, 8],
            # Columns
            [0, 3, 6], [1, 4, 7], [2, 5, 8],
            # Diagonals
            [0, 4, 8], [2, 4, 6],
        ]

        for line in lines:
            if all(self.board[i] == mark for i in line):
                return True
        return False

    def _is_board_full(self) -> bool:
        """Check if all cells are occupied."""
        return all(cell != "" for cell in self.board)

    def to_dict(self) -> dict:
        """Convert game state to dictionary for JSON serialization."""
        result = {
            "id": self.id,
            "board": self.board,
            "currentPlayer": self.current_player.value,
            "status": self.status.value,
        }
        if self.winner:
            result["winner"] = self.winner.value
        return result


class GameStore:
    """Manages game instances."""

    def __init__(self):
        self._games: dict[str, Game] = {}

    def create(self) -> Game:
        """Create a new game and store it."""
        game = Game()
        self._games[game.id] = game
        return game

    def get(self, game_id: str) -> Optional[Game]:
        """Retrieve a game by ID."""
        return self._games.get(game_id)

File: apps/go-tictactoe/game/game.go
package game

import (
	"errors"
	"sync"

	"github.com/google/uuid"
)

// Player represents X or O
type Player string

const (
	PlayerX Player = "X"
	PlayerO Player = "O"
)

// Status represents the game state
type Status string

const (
	StatusInProgress Status = "in_progress"
	StatusXWins      Status = "x_wins"
	StatusOWins      Status = "o_wins"
	StatusDraw       Status = "draw"
	StatusXForfeits  Status = "x_forfeits"
	StatusOForfeits  Status = "o_forfeits"
)

// Error types for invalid operations
var (
	ErrNotYourTurn     = errors.New("not_your_turn")
	ErrCellOccupied    = errors.New("cell_occupied")
	ErrInvalidPosition = errors.New("invalid_position")
	ErrGameOver        = errors.New("game_over")
)

// Game represents a tic-tac-toe game
type Game struct {
	ID            string   `json:"id"`
	Board         [9]string `json:"board"`
	CurrentPlayer Player   `json:"currentPlayer"`
	Status        Status   `json:"status"`
	Winner        *Player  `json:"winner,omitempty"`
}

// New creates a new game with X as the first player
func New() *Game {
	return &Game{
		ID:            uuid.New().String(),
		Board:         [9]string{"", "", "", "", "", "", "", "", ""},
		CurrentPlayer: PlayerX,
		Status:        StatusInProgress,
	}
}

// Move places a mark on the board
func (g *Game) Move(row, col int, player Player) error {
	// Check if game is over
	if g.Status != StatusInProgress {
		return ErrGameOver
	}

	// Validate position
	if row < 0 || row > 2 || col < 0 || col > 2 {
		return ErrInvalidPosition
	}

	// Check turn
	if player != g.CurrentPlayer {
		return ErrNotYourTurn
	}

	// Calculate board index
	index := row*3 + col

	// Check if cell is occupied
	if g.Board[index] != "" {
		return ErrCellOccupied
	}

	// Place the mark
	g.Board[index] = string(player)

	// Check for win
	if g.checkWin(player) {
		if player == PlayerX {
			g.Status = StatusXWins
		} else {
			g.Status = StatusOWins
		}
		g.Winner = &player
		return nil
	}

	// Check for draw
	if g.isBoardFull() {
		g.Status = StatusDraw
		return nil
	}

	// Switch player
	if g.CurrentPlayer == PlayerX {
		g.CurrentPlayer = PlayerO
	} else {
		g.CurrentPlayer = PlayerX
	}

	return nil
}

// Forfeit allows the current player to forfeit the game
func (g *Game) Forfeit(player Player) error {
	// Check if game is over
	if g.Status != StatusInProgress {
		return ErrGameOver
	}

	// Check turn
	if player != g.CurrentPlayer {
		return ErrNotYourTurn
	}

	// Set forfeit status and winner
	if player == PlayerX {
		g.Status = StatusXForfeits
		opponent := PlayerO
		g.Winner = &opponent
	} else {
		g.Status = StatusOForfeits
		opponent := PlayerX
		g.Winner = &opponent
	}

	return nil
}

// checkWin checks if the given player has won
func (g *Game) checkWin(player Player) bool {
	mark := string(player)

	// Winning combinations (rows, columns, diagonals)
	lines := [][3]int{
		// Rows
		{0, 1, 2}, {3, 4, 5}, {6, 7, 8},
		// Columns
		{0, 3, 6}, {1, 4, 7}, {2, 5, 8},
		// Diagonals
		{0, 4, 8}, {2, 4, 6},
	}

	for _, line := range lines {
		if g.Board[line[0]] == mark &&
			g.Board[line[1]] == mark &&
			g.Board[line[2]] == mark {
			return true
		}
	}
	return false
}

// isBoardFull checks if all cells are occupied
func (g *Game) isBoardFull() bool {
	for _, cell := range g.Board {
		if cell == "" {
			return false
		}
	}
	return true
}

// BoardSlice returns board as a slice for JSON serialization
func (g *Game) BoardSlice() []string {
	return g.Board[:]
}

// Store manages game instances
type Store struct {
	games map[string]*Game
	mu    sync.RWMutex
}

// NewStore creates a new game store
func NewStore() *Store {
	return &Store{
		games: make(map[string]*Game),
	}
}

// Create creates a new game and stores it
func (s *Store) Create() *Game {
	s.mu.Lock()
	defer s.mu.Unlock()

	game := New()
	s.games[game.ID] = game
	return game
}

// Get retrieves a game by ID
func (s *Store) Get(id string) (*Game, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	game, ok := s.games[id]
	return game, ok
}
